src/Datadog.AutoInstrumentation.NativeLoader/string.cpp

```diff
@@ -0,0 +1,181 @@
+    return miniutf::to_utf8(ustr);
+#endif
+}
+std::string ToString(const LPTSTR& tstr)
+{
+#if defined(_UNICODE)
+    return ToString(WSTRING(tstr));
+#else
+    return std::string(tstr);
+#endif
+}
+
+WSTRING ToWSTRING(const std::string& str)
+{
+#ifdef _WIN32
+    if (str.empty()) return std::wstring();
+
+    std::wstring tmpStr(tmp_buffer_size, 0);
+    int size_needed = MultiByteToWideChar(CP_UTF8, 0, &str[0], (int) str.size(), &tmpStr[0], tmp_buffer_size);
+    if (size_needed < tmp_buffer_size)
+    {
+        return tmpStr.substr(0, size_needed);
+    }
+
+    std::wstring wstrTo(size_needed, 0);
+    MultiByteToWideChar(CP_UTF8, 0, &str[0], (int) str.size(), &wstrTo[0], size_needed);
+    return wstrTo;
+#else
+    auto ustr = miniutf::to_utf16(str);
+    return WSTRING(reinterpret_cast<const WCHAR*>(ustr.c_str()));
+#endif
+}
+
+WSTRING ToWSTRING(const uint64_t i)
+{
+    return WSTRING(reinterpret_cast<const WCHAR*>(std::to_wstring(i).c_str()));
+}
+
+constexpr char HexMap[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+
+WSTRING HexStr(const void* dataPtr, int len)
+{
+    const unsigned char* data = (unsigned char*) dataPtr;
+    WSTRING s(len * 2, ' ');
+    for (int i = 0; i < len; ++i)
+    {
+        s[2 * i] = HexMap[(data[i] & 0xF0) >> 4];
+        s[2 * i + 1] = HexMap[data[i] & 0x0F];
+    }
+    return s;
+}
+
+WSTRING Trim(const WSTRING& str)
+{
+    if (str.length() == 0)
+    {
+        return WStr("");
+    }
+
+    WSTRING trimmed = str;
+
+    auto lpos = trimmed.find_first_not_of(WStr(" \t"));
+    if (lpos != WSTRING::npos && lpos > 0)
+    {
+        trimmed = trimmed.substr(lpos);
+    }
+
+    auto rpos = trimmed.find_last_not_of(WStr(" \t"));
+    if (rpos != WSTRING::npos)
+    {
+        trimmed = trimmed.substr(0, rpos + 1);
+    }
+
+    return trimmed;
+}
+
+std::string Trim(const std::string& str)
+{
+    if (str.length() == 0)
+    {
+        return "";
+    }
+
+    std::string trimmed = str;
+
+    auto lpos = trimmed.find_first_not_of(" \t");
+    if (lpos != std::string::npos && lpos > 0)
+    {
+        trimmed = trimmed.substr(lpos);
+    }
+
+    auto rpos = trimmed.find_last_not_of(" \t");
+    if (rpos != std::string::npos)
+    {
+        trimmed = trimmed.substr(0, rpos + 1);
+    }
+
+    return trimmed;
+}
+
+template <typename Out>
+void Split(const WSTRING& s, wchar_t delim, Out result)
+{
+    size_t lpos = 0;
+    for (size_t i = 0; i < s.length(); i++)
+    {
+        if (s[i] == delim)
+        {
+            *(result++) = s.substr(lpos, (i - lpos));
+            lpos = i + 1;
+        }
+    }
+    *(result++) = s.substr(lpos);
+}
+
+std::vector<WSTRING> Split(const WSTRING& s, wchar_t delim)
+{
+    std::vector<WSTRING> elems;
+    Split(s, delim, std::back_inserter(elems));
+    return elems;
+}
+
+template <typename Out>
+void Split(const std::string& s, char delim, Out result)
```

Comment:
Perhaps I am not see-plus-plusy enough, but would it not be simpler and robuster to simply use the vector adding logic in the vector overload and not have this fancy templated overload? (this is just for my own education)

Reply:
`string.cpp` was a direct copy from the tracer project, I think we can revisit those function contained in that file in other PR. The tracer will benefit from any optimization here.

Reply:
❗ [follow up in future PR]